x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleRunner.java

```diff
@@ -294,41 +301,48 @@ ClusterState moveClusterStateToFailedStep(ClusterState currentState, String[] in
         return newState;
     }
 
-    private static Settings.Builder moveIndexSettingsToNextStep(LifecyclePolicy policy, Settings existingSettings,
-                                                                StepKey currentStep, StepKey nextStep, LongSupplier nowSupplier) {
+    private static LifecycleExecutionState moveExecutionStateToNextStep(LifecyclePolicy policy,
+                                                                        LifecycleExecutionState existingState,
+                                                                        StepKey currentStep, StepKey nextStep,
+                                                                        LongSupplier nowSupplier) {
         long nowAsMillis = nowSupplier.getAsLong();
-        Settings.Builder newSettings = Settings.builder().put(existingSettings).put(LifecycleSettings.LIFECYCLE_PHASE, nextStep.getPhase())
-                .put(LifecycleSettings.LIFECYCLE_ACTION, nextStep.getAction()).put(LifecycleSettings.LIFECYCLE_STEP, nextStep.getName())
-                .put(LifecycleSettings.LIFECYCLE_STEP_TIME, nowAsMillis)
-                // clear any step info or error-related settings from the current step
-                .put(LifecycleSettings.LIFECYCLE_FAILED_STEP, (String) null)
-                .put(LifecycleSettings.LIFECYCLE_STEP_INFO, (String) null);
+        LifecycleExecutionState.Builder updatedState = LifecycleExecutionState.builder(existingState);
+        updatedState.setPhase(nextStep.getPhase());
+        updatedState.setAction(nextStep.getAction());
+        updatedState.setStep(nextStep.getName());
+        updatedState.setStepTime(nowAsMillis);
+
+        // clear any step info or error-related settings from the current step
+        updatedState.setFailedStep("");
+        updatedState.setStepInfo("");
```

Comment:
I think we should set these to `null` so they are subsequently removed from the map rather than use an empty string